File: verify.js
"use strict";

const fs = require("node:fs");
const path = require("node:path");
const Sutl = require("./reference.js");

const root = __dirname;
const corpus = JSON.parse(fs.readFileSync(path.join(root, "conformance.json"), "utf8"));

function same(left, right) {
  if (typeof left !== typeof right) return false;
  if (left === null || right === null) return left === right;
  if (Array.isArray(left) || Array.isArray(right))
    return (
      Array.isArray(left) &&
      Array.isArray(right) &&
      left.length === right.length &&
      left.every((value, index) => same(value, right[index]))
    );
  if (typeof left === "object") {
    const leftKeys = Object.keys(left).sort();
    const rightKeys = Object.keys(right).sort();
    return (
      same(leftKeys, rightKeys) &&
      leftKeys.every((key) => same(left[key], right[key]))
    );
  }
  return Object.is(left, right);
}

function bundleLibrary(bundle) {
  const declarations = bundle.library || [];
  const library = {};
  for (const declaration of declarations)
    if (declaration?.name && "transform-t" in declaration)
      library[declaration.name] = declaration["transform-t"];
  for (const required of bundle.declaration?.requires || []) {
    const match = declarations.find((item) => (item.name || "").startsWith(required));
    if (match) library[required] = match["transform-t"];
  }
  return library;
}

function runCase(testCase) {
  const source = testCase.source;
  const library = { ...(testCase.library || {}) };
  const mode = testCase.mode || "evaluate";
  if (mode === "evaluate")
    return Sutl.evaluate(source, testCase.transform, library);
  if (mode === "compilelib_evaluate") {
    const compiled = Sutl.compilelib(
      [testCase.declaration],
      testCase.distributions || [],
      library,
      Boolean(testCase.test),
    );
    if (compiled.fail) return { "compile-fail": compiled.fail };
    return Sutl.evaluate(source, testCase.declaration["transform-t"], compiled.lib);
  }
  if (mode === "declaration_test_t")
    return !Sutl.compilelib(
      [testCase.declaration],
      testCase.distributions || [],
      library,
      true,
    ).fail;
  if (mode === "studio_fixture_set") {
    const directory = path.join(root, path.dirname(testCase.fixture_glob));
    const files = fs.readdirSync(directory).filter((name) => name.endsWith(".json")).sort();
    const failures = [];
    for (const name of files) {
      const fixture = JSON.parse(fs.readFileSync(path.join(directory, name), "utf8"));
      try {
        const actual = Sutl.evaluate(
          fixture.source,
          fixture.transform,
          bundleLibrary(fixture),
        );
        if (!same(actual, fixture.expected)) failures.push(fixture.id);
      } catch (error) {
        failures.push(`${fixture.id}: ${error.message}`);
      }
    }
    return { evaluated: files.length, failures };
  }
  throw new Error(`unknown conformance mode: ${mode}`);
}

const results = corpus.cases.map((testCase) => ({
  id: testCase.id,
  actual: runCase(testCase),
}));

if (process.argv.includes("--json")) {
  process.stdout.write(`${JSON.stringify(results)}\n`);
  process.exit(0);
}

const failures = [];
for (const result of results) {
  const testCase = corpus.cases.find((item) => item.id === result.id);
  const actual = result.actual;
  if (!same(actual, testCase.expected))
    failures.push([testCase.id, testCase.expected, actual]);
}
for (const [id, expected, actual] of failures) {
  console.log(`FAIL ${id}`);
  console.log(`  expected: ${JSON.stringify(expected)}`);
  console.log(`  actual:   ${JSON.stringify(actual)}`);
}
console.log(`${corpus.cases.length - failures.length}/${corpus.cases.length} passed`);
process.exitCode = failures.length ? 1 : 0;

File: verify.py
"""Run the normative sUTL 1.0 corpus against the Python reference."""
from __future__ import annotations

import json
from pathlib import Path
from typing import Any

from reference import compilelib, evaluate


ROOT = Path(__file__).resolve().parent
CORPUS = json.loads((ROOT / "conformance.json").read_text(encoding="utf-8"))


def same(left: Any, right: Any) -> bool:
    if isinstance(left, bool) or isinstance(right, bool):
        return type(left) is type(right) and left == right
    if isinstance(left, (int, float)) and isinstance(right, (int, float)):
        return left == right
    if type(left) is not type(right):
        return False
    if isinstance(left, dict):
        return left.keys() == right.keys() and all(
            same(left[key], right[key]) for key in left
        )
    if isinstance(left, list):
        return len(left) == len(right) and all(
            same(a, b) for a, b in zip(left, right)
        )
    return left == right


def bundle_library(bundle: dict[str, Any]) -> dict[str, Any]:
    declarations = bundle.get("library") or []
    library = {
        declaration["name"]: declaration["transform-t"]
        for declaration in declarations
        if isinstance(declaration, dict)
        and "name" in declaration
        and "transform-t" in declaration
    }
    for required in (bundle.get("declaration") or {}).get("requires", []):
        match = next(
            (
                declaration
                for declaration in declarations
                if declaration.get("name", "").startswith(required)
            ),
            None,
        )
        if match:
            library[required] = match["transform-t"]
    return library


def run_case(case: dict[str, Any]) -> Any:
    source = case.get("source")
    library = dict(case.get("library", {}))
    mode = case.get("mode", "evaluate")
    if mode == "evaluate":
        return evaluate(source, case["transform"], library)
    if mode == "compilelib_evaluate":
        compiled = compilelib(
            [case["declaration"]],
            case.get("distributions", []),
            seed=library,
            test=case.get("test", False),
        )
        if "fail" in compiled:
            return {"compile-fail": compiled["fail"]}
        return evaluate(source, case["declaration"]["transform-t"], compiled["lib"])
    if mode == "declaration_test_t":
        return "fail" not in compilelib(
            [case["declaration"]],
            case.get("distributions", []),
            seed=library,
            test=True,
        )
    if mode == "studio_fixture_set":
        failures = []
        paths = sorted(ROOT.glob(case["fixture_glob"]))
        for path in paths:
            fixture = json.loads(path.read_text(encoding="utf-8"))
            try:
                actual = evaluate(
                    fixture.get("source"),
                    fixture.get("transform"),
                    bundle_library(fixture),
                )
                if not same(actual, fixture.get("expected")):
                    failures.append(fixture["id"])
            except Exception as error:  # pragma: no cover - surfaced in output
                failures.append(f"{fixture['id']}: {error}")
        return {"evaluated": len(paths), "failures": failures}
    raise ValueError(f"unknown conformance mode: {mode}")


def main() -> None:
    failed = []
    for case in CORPUS["cases"]:
        actual = run_case(case)
        if not same(actual, case["expected"]):
            failed.append((case["id"], case["expected"], actual))
    for case_id, expected, actual in failed:
        print(f"FAIL {case_id}")
        print(f"  expected: {json.dumps(expected, ensure_ascii=False)}")
        print(f"  actual:   {json.dumps(actual, ensure_ascii=False)}")
    print(f"{len(CORPUS['cases']) - len(failed)}/{len(CORPUS['cases'])} passed")
    raise SystemExit(1 if failed else 0)


if __name__ == "__main__":
    main()
